Lib/functools.py

```diff
@@ -273,33 +273,100 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
+
+class PlaceholderType:
+    """PlaceholderType()
+    --
+
+    The type of the Placeholder singleton.
+    Used as a placeholder for partial arguments.
+    """
+    _singleton = None
+
+    def __new__(cls):
+        if cls._singleton is None:
+            cls._singleton = object.__new__(cls)
+        return cls._singleton
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __bool__(self):
+        raise TypeError("Placeholder should not be used in a boolean context")
+
+    def __reduce__(self):
+        return type(self), ()
+
+
+Placeholder = PlaceholderType()
+
+
 # Purely functional, no descriptor behaviour
 class partial:
     """New function with partial application of the given arguments
     and keywords.
     """
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    __slots__ = ("func", "args", "keywords", "placeholder_count",
+                 "__dict__", "__weakref__")
 
     def __new__(cls, func, /, *args, **keywords):
         if not callable(func):
             raise TypeError("the first argument must be callable")
-
+        placeholder_count = 0
+        nargs = len(args)
+        if args:
+            while nargs and args[nargs-1] is Placeholder:
+                nargs -= 1
+            args = args[:nargs]
+            placeholder_count = args.count(Placeholder)
         if isinstance(func, partial):
-            args = func.args + args
+            pargs = func.args
+            pnp = func.placeholder_count
+            if pnp and args:
+                all_args = list(pargs)
+                nargs = len(args)
+                j, pos = 0, 0
+                end = nargs if nargs < pnp else pnp
+                while j < end:
+                    pos = all_args.index(Placeholder, pos)
+                    all_args[pos] = args[j]
+                    j += 1
+                    pos += 1
+                if pnp < nargs:
+                    all_args.extend(args[pnp:])
+                placeholder_count += pnp - end
+                args = tuple(all_args)
+            else:
+                placeholder_count += pnp
+                args = func.args + args
             keywords = {**func.keywords, **keywords}
             func = func.func
 
         self = super(partial, cls).__new__(cls)
-
         self.func = func
         self.args = args
         self.keywords = keywords
+        self.placeholder_count = placeholder_count
         return self
 
     def __call__(self, /, *args, **keywords):
         keywords = {**self.keywords, **keywords}
-        return self.func(*self.args, *args, **keywords)
+        if placeholder_count := self.placeholder_count:
+            if len(args) < placeholder_count:
```

Comment:
How do you deal with partial functions with *default* values? I think we should allow constructions of the form:

```py
def foo(a=1, b=2, c=3):
	return (a, b, c)

partial_foo = partial(foo, PLACEHOLDER, PLACEHOLDER)
assert partial_foo() == (1, 2, 3)
assert partial_foo(0) == (0, 2, 3)
```

Reply:
Currently, the first step is that all trailing `placeholders` are trimmed. So:
```
partial_foo = partial(foo, PLACEHOLDER, PLACEHOLDER)
```
is the same as
```
partial_foo = partial(foo)
```
However, now I am thinking to raise an exception and disallow trailing placeholders. I think it will eliminate a certain amount of confusion and will make initialiser a bit faster (especially Python version).